configfile: "config/main.yaml"

rule create_barcodes:
    input:
        singlebams="data/aligned/" + str(config["SLX"]) + "/{sample}.bam"
    output:
        "results/sample_barcodes_" + str(config["SLX"]) + ".tsv"
    shell:
        """
        python makebarc.py -i {input} -o {output}
        """

rule mark_and_merge_bams:
    input:
        singlebams="data/aligned/" + str(config["SLX"]) + "/{sample}.bam"
        samplesheet="sample_barcodes_" + str(config["SLX"]) + ".tsv"
        refsheet="num_to_chr.tsv"
    output:
        "results/merged_bams/" + str(config["SLX"]) + "/{sample}_merged.bam"
    shell:
        """
        sbatch /mnt/scratchc/fmlab/kasper01/helper_tools/mark_and_merge_bams_samplesheet \
        -i {input.singlebams} -s {input.samplesheet} -r {input.refsheet}
        """
    
rule sort_and_index:
    input:
        rules.mark_and_merge_bams.output
    output:
        bam="results/merged_bams/" + str(config["SLX"]) + "/{sample}_merged_sorted.bam"
        bai="results/merged_bams/" + str(config["SLX"]) + "/{sample}_merged_sorted.bam.bai"
    threads:
        15
    shell:
        """
        samtools sort {input} -o {output.bam}
        samtools index {output.bam} -o {output.bai}
        """

rule chisel_rdr:
    input:
        mergedbam="results/merged_bams/" + str(config["SLX"]) + "/{sample}_merged_sorted.bam"
        refgenome=config["path_ref_genome"]
    output:
        "results/chisel_rdr/" + str(config["SLX"]) + "/"
    shell:
        """
        chisel_rdr -t {input.merged_bams} -r {input.refgenome} -b 50000 -m 100000 -j 15
        """

rule run_sprinter:
    input:
        "results/chisel_rdr/" + str(config["SLX"]) + "/rdr/rdr.tsv"
        refgenome=config["path_ref_genome"]
    output:
        tsv="results/chisel_rdr/" + str(config["SLX"]) + "/sprinter_{SLX}.input.tsv.gz"
        result="results/sprinter_result/" + str(config["SLX"]) + "/"
    shell:
        """
        gzip -c {input} > {output.tsv}
        rm -rf rdr/ baf/ combo/ calls/ clones/ plots/

        sprinter {output.tsv} \
         --refgenome {input.refgenome} \
         --minreads 100000 \
         --rtreads 200 \
         --cnreads 1000 \
         --minnumcells 15 \
         -j 12
        """